### PR TITLE
Implement Tensor.new_empty_strided(sizes, strides, *, dtype, device, requires_grad)

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #47227 Batched gradient support for view+inplace operations
* #47365 Add batching rule for torch.clone(tensor, torch.contiguous_format)
* #47364 as_strided batching rule
* #47226 Batching rule for Tensor.new_empty_strided
* **#47225 Implement Tensor.new_empty_strided(sizes, strides, *, dtype, device, requires_grad)**

Summary
-------
This PR implements Tensor.new_empty_strided. Many of our torch.* factory
functions have a corresponding new_* method (e.g., torch.empty and
torch.new_empty), but there is no corresponding method to
torch.empty_strided. This PR adds one.

Motivation
----------
The real motivation behind this is for vmap to be able to work through
CopySlices. CopySlices shows up a lot in double backwards because a lot
of view functions have backward formulas that perform view+inplace.

https://github.com/pytorch/pytorch/blob/e0fd590ec950cb1e65ea0431c9e765f8cda27908/torch/csrc/autograd/functions/tensor.cpp#L78-L106

To support vmap through CopySlices, the approach in this stack is to:
- add `Tensor.new_empty_strided` and replace `empty_strided` in
CopySlices with that so that we can propagate batch information.
- Make some slight modifications to AsStridedBackward (and add
as_strided batching rule)

Please let me know if it would be better if I squashed everything related to
supporting vmap over CopySlices together into a single big PR.

Test Plan
---------
- New tests.

Differential Revision: [D24741688](https://our.internmc.facebook.com/intern/diff/D24741688)

### CODE REVIEW

File: aten/src/ATen/native/TensorFactories.cpp [hunk 1]
```diff
@@ -341,6 +341,15 @@ Tensor new_empty(
   return at::empty(size, self.options().merge_in(options));
 }
 
+Tensor new_empty_strided(
+    const Tensor& self,
+    IntArrayRef size,
+    IntArrayRef strides,
```
Comment (albanD) [inline aten/src/ATen/native/TensorFactories.cpp:347]: nit: size is singular but stride is plural in all the signature. Should we make them consistent?

**Review by albanD — APPROVED** [review]
lgtm

File: aten/src/ATen/native/TensorFactories.cpp (hunk 1 above)
Comment (zou3519) [reply aten/src/ATen/native/TensorFactories.cpp:347]: Changed to singular to match existing the signatures of existing strided functions (as_strided, empty_strided)

**Comment by albanD** [comment]
Note that CI failures are real because `__torch_function__` is not properly handled by the current implementation.

**Review by ezyang — APPROVED** [review]
Separate is better. Thanks!
